### ISSUE TITLE
Ability to backup to more destination types

### ISSUE THREAD

**Issue by saidul** [Dokploy/dokploy#416]
### What problem will this feature address?

Can not backup to FTP, SFTP, Google Drive other storage solutions.

### Describe the solution you'd like

Add support for other destination types like Google Drive, One Drive, FTP, SFTP etc.

### Describe alternatives you've considered

During backup destination creation, we can select destination types and their credentials accordingly. This can later be used in upload phase of backup process. Following can be used for uploading the data:
- rclone - supports a lot of destination. destination configuration can be given as argument. Using rclone will probably be easiest and fastest to add support for many providers.
- https://github.com/duna-oss/flystorage - common api for services. adapter based. 
- Independent libraries - will require more effort but provides granular control over each api.

**Comment by Siumauricio** [comment]
Sounds like a good idea, it is in my plans to come up with a way to support multiple destination sources also for backups and more, I think rclone looks like a very good suggestion as it supports many ways, I will add to the roadmap

**Comment by Siumauricio** [comment]
Partially closed by #469, we need to adjust a little bit the code in order to support more providers, but we are using RCLONE for now for S3

**Comment by kerimovok** [comment]
Hello, are there any updates on this?

**Comment by saidul** [comment]
@Siumauricio Thank you so much for adding rclone support. However we still can't backup to any provider other than S3.

Rclone supports configuration via:
- Environment variables: https://rclone.org/docs/#config-file
- Connection strings: https://rclone.org/docs/#connection-strings
- Config file

Different providers require different parameters. If we can support passing either environment variables or a connection string to the rclone command, it would enable compatibility with a wider range of storage providers.

I imagine following dialog can be adjusted to support advanced mode.  When advanced mode/custom mode is selected, other fields related to s3 will be hidden and only 1 textbox will be present to take either the connection string or the environment variables. 


![Image](https://github.com/user-attachments/assets/bb8f9009-110c-438a-b1fa-e2c6b065f81a)

My humble opinion is that the environment variable is easier to define. Because when json data needs to be provided in connection string, it becomes complicated to escape the string.  

Some provider requires connection info to be updated time to time. For example in case of google drive, rclone automatically refresh the tokens and store that in the config file. In such case it is probably better to have a config file defined in some volume for dockploy (like traefik config files can be modified by users seprately).

So for each backup destination we can have different config files. Which can be passed to rclone during execution of the command.

**Comment by maelp** [comment]
Would be nice if we could also drag-and-drop a .zip file in the interface to restore from that backup, as an "escape hatch", and also save a manual backup as a .zip file from the interface

**Comment by AlexMason** [comment]
RE: Algora.

This is amazing. I love incentivized rewards for contributing. I work a lot in my day job, and it's hard to justify contributing to opensource (even projects I use). I hope this can make contributing more accessible to those who have to choose between food on the table, and support the literal backbone of their work. (please allow replies on newsposts in discord, I don't like using issues as a forum for these comments).

**Comment by saidul** [comment]
Great work on your PR @vishalkadam47 

Could you please also consider following things for rclone?

1. Support [Crypt](https://rclone.org/crypt/) This basically encrypts other storage systems making them E2EE.
2. Support for custom rclone configuration snippet - This will keep the opportunity open to use any storage backend supported by rclone.

**Comment by vishalkadam47** [comment]
> Great work on your PR @vishalkadam47 
> 
> Could you please also consider following things for rclone?
> 
> 1. Support [Crypt](https://rclone.org/crypt/) This basically encrypts other storage systems making them E2EE.
> 2. Support for custom rclone configuration snippet - This will keep the opportunity open to use any storage backend supported by rclone. 
> 


Thanks, sure will take a look at Crypt and add support for more storage providers.
Waiting for @Siumauricio review on https://github.com/Dokploy/dokploy/pull/1979

**Comment by raqibnur** [comment]
@vishalkadam47 Impressive attempt — very near completion. When can we see this in production?

**Comment by lweyrich1** [comment]
hey - i can take this bounty if it's still active. i'll aim for an MVP using rclone as suggested (FTP/SFTP/GDrive/OneDrive etc) with minimal UI/config changes and a safe creds flow.\n\ncan you confirm current acceptance criteria (which providers are must-have for the ) + whether there's anything to do on Algora to claim/assign?

**Comment by teredasites** [comment]
C:/Program Files/Git/attempt #416

**Comment by yuliuyi717-ux** [comment]
/attempt #416

Plan:
1) Map current backup pipeline and provider abstraction points (storage adapters, credential model, scheduling flow).
2) Add one incremental destination backend first (S3-compatible baseline) with clear interface so additional providers can be plugged in without touching core backup logic.
3) Wire provider-specific validation + config persistence, and keep current backup behavior unchanged for existing destinations.
4) Add regression tests for destination selection, upload path, and error handling so failures surface with actionable logs.

I will keep this as a small, reviewable PR and then extend to additional destination types if maintainers want it in follow-up commits.

**Comment by yuliuyi717-ux** [comment]
Implemented and submitted for review: https://github.com/Dokploy/dokploy/pull/3874

Scope:
- provider-aware rclone destination support across backup flows (S3 + FTP + SFTP + Google Drive + OneDrive)
- provider-aware test connection
- destination UI/provider options update
- utility tests for provider mapping and destination path generation

**Comment by justinleeyang** [comment]
I can take bounty #416. I will share a scoped implementation plan and first PR shortly.

**Comment by justinleeyang** [comment]
Quick follow-up on #416: I am still actively working on this. Please let me know if you want me to narrow scope to a smaller first PR.

**Comment by justinleeyang** [comment]
Quick follow-up on #416: I am still actively working on this. Please let me know if you want me to narrow scope to a smaller first PR.

**Comment by justinleeyang** [comment]
Quick follow-up on #416: I am still actively working on this. Please let me know if you want me to narrow scope to a smaller first PR.

**Comment by justinleeyang** [comment]
Quick follow-up on #416: I am still actively working on this. Please let me know if you want me to narrow scope to a smaller first PR.

**Comment by WuXieSec** [comment]
I’m looking into this and plan to submit a PR shortly. I’ll scope a clean extension path for additional backup destinations (starting with SFTP/FTP) and keep backward compatibility.

**Comment by edsonalanis** [comment]
Hello, I'm very interested in the backup option on Google Drive. I'd like to know the status of its development and if there's any estimated release date.

**Comment by marcoberutti** [comment]
hello you are asking to me? thanks

**Comment by jaysu66** [comment]
C:/Program Files (x86)/Git/attempt #416

**Comment by heickerv1001-dev** [comment]
/claim #416

Working on generic rclone remote support for backup destinations.

**Comment by edsonalanis** [comment]
Hello, are there any updates on this?

**Comment by jhhjwei** [comment]
/attempt #416

I’m proceeding with a focused, staged implementation for generic rclone-backed backup destinations while preserving the current S3 flow:

1. Introduce a destination type/config abstraction without breaking existing S3 records.
2. Add a generic rclone remote/config path for Google Drive, OneDrive, FTP and SFTP.
3. Reuse the abstraction for connection testing, upload and retention cleanup.
4. Add validation and tests for command construction and backward compatibility.
5. Keep the first PR reviewable; split UI/migration work if needed.

I’ve completed the code-path analysis and am preparing the implementation branch now.

**Comment by edmogeor** [comment]
As a workaround for now until the changes are merged, if anyone wants to use rclone... 

1. Generate an s3 access key and secret key:

```
openssl rand -hex 16  # S3 access key
openssl rand -hex 32  # S3 secret key`
```

2. Setup a compose file in dokploy with this, change example-remote and example-path and put an S3 access key and secret key in an env:

```yaml
services:
  rclone-s3:
    image: rclone/rclone:1.71.0
    restart: unless-stopped
    networks:
      - dokploy-network
    command:
      - serve
      - s3
      - "example-remote:example-path"
      - --addr
      - ":8080"
      - --auth-key
      - ${BACKUP_S3_ACCESS_KEY},${BACKUP_S3_SECRET_KEY}
      - --server-read-timeout
      - 2h
      - --server-write-timeout
      - 2h
      - --force-path-style=true
    environment:
      RCLONE_CONFIG: /config/rclone.conf
    volumes:
      - ../files/rclone.conf:/config/rclone.conf:ro

networks:
  dokploy-network:
    external: true
```

3. Use a trusted machine to create the rclone config, then create a file mount in advanced settings with the path "rclone.conf", use the same remote name as you used above...

to create the config:
`rclone config`

to find the file path:
`rclone config file`

4. Then you can just setup an S3 Destination in Dokploy as usual and point it to http://rclone-s3:8080

Not ideal and as easy to manage, but it works!

**Comment by Prodgs** [comment]
Hi! I’d like to work on this bounty, and I do want the bounty if the fix is accepted. I’m reviewing the current issue and repository now. If the issue description reflects the current expected behavior, I’ll work against that; if there are additional acceptance criteria or bounty-specific requirements, please let me know. Thanks!

**Update:** After reviewing the existing implementation attempts and current competition, I’m withdrawing from this bounty for now. Thanks!

**Comment by GAB-333** [comment]
The Algora board currently lists this issue as an open $50 bounty despite several prior claims. Before starting another implementation: is the bounty still funded and active, and which backup-destination type or acceptance gap remains unimplemented by the existing PRs? I can take a bounded missing destination once the current payout and merge conditions are confirmed.

**Comment by mobiusbbbot-lab** [comment]
/attempt #416

I’d like to take a focused remaining slice of this Algora $50 bounty (TypeScript / Dokploy backup destinations).

Plan:
1. Audit current `main` + open PRs (#1979, #4208, #4546, #4563, #4918) to identify what is still missing vs the issue ask (generic rclone / additional destination types beyond what’s already mergeable).
2. Implement only a non-overlapping gap (e.g. one missing destination type + credential UI + backup job wiring + tests), not a duplicate of the generic-rclone PRs already open.
3. Keep the PR scoped, with docs for configuring the new destination and a smoke test path.
4. Open PR with `/claim #416`.

Before coding the full slice: please confirm the bounty is still payable given the open rclone PRs (same question @GAB-333 raised). If a specific destination is preferred, I’ll target that first.

ETA: this week once the remaining gap is confirmed.

**Comment by Furox-Art** [comment]
Thanks for the positive feedback on the idea!

**Comment by gbert64-lab** [comment]
/attempt #416

Plan (gbert64-lab / Cursor): map existing backup destinations + open PRs (#1979, #4208, #4546, #4563, #4918, #5349). Implement ONE remaining gap only (prefer a clean rclone slice not already covered), with tests. Open PR with `/claim #416`. Will kill if gap already closed by open PRs.

**Comment by jr92yhnt8b-commits** [comment]
Algora currently lists this as an open $50 bounty, but this thread has several prior attempts/PRs and the original scope has partially changed since rclone landed. Before I duplicate work, can the maintainer/bounty sponsor confirm that the $50 bounty is still active and specify the remaining acceptance scope that is eligible for payout? If confirmed, I can take a focused unimplemented provider/slice rather than overlap existing work.

**Comment by pranavvehmid** [comment]
Hi! I’m interested in contributing an AI-assisted fix. Before starting, could you confirm whether the $50 bounty is still available and what work remains beyond the existing PRs? Are there any project rules for AI-assisted contributions?

**Comment by woahwhattheheck** [comment]
Current-main implementation note for #416 after tracing the existing rclone path:

The smallest non-duplicative route is to generalize the **existing Destination** abstraction rather than introduce a parallel cloud-backup schema/router stack.

Current main already has a provider-agnostic upload primitive: `getBackupCommand(backup, rcloneFlags, rcloneDestination, logPath)`. The S3 coupling is duplicated around it in four places:
1. `apps/dokploy/server/api/routers/destination.ts::testConnection` hardcodes S3 flags + `:s3:${bucket}`.
2. `packages/server/src/utils/backups/*` (for example postgres.ts) builds `:s3:${destination.bucket}/...`.
3. `keepLatestNBackups()` hardcodes `:s3:${destination.bucket}`.
4. `backupRouter.listBackupFiles` and the restore utilities (for example `utils/restore/postgres.ts`) hardcode the same S3 credential/path construction.

A bounded implementation can add a destination discriminator (existing S3 vs generic rclone remote/config) and one helper that returns something like `{ flags, remoteRoot }`; then route **test, upload, retention, list, and restore** through that helper. Existing S3 rows continue to produce the current flags/root unchanged. Generic rclone destinations can then represent Drive/OneDrive/FTP/SFTP through a validated config/remote rather than requiring a second backup table and second scheduler.

Acceptance tests I would require before claiming the bounty:
- legacy S3 helper output is byte-for-byte equivalent;
- one generic remote fixture exercises test + upload + retention + list + restore path construction;
- malformed remote/config input fails closed (no shell interpolation);
- secrets in generic rclone config/flags are covered by redaction, not just the two current S3 flag regexes;
- Google Drive token refresh has a persistent config location (an ephemeral command-line config is insufficient).

This also avoids reviving PR #1979's separate `cloudStorageBackup` / `cloudStorageDestination` architecture, which is still open/non-mergeable and whose added cloud schema/service/router files are absent from literal current main.

**Comment by tokenjunkielabs** [comment]
/attempt #416

I’m taking the current-main rclone generalization path: preserve existing S3 behavior, add a validated generic remote/config destination, and route connection test, backup upload, retention/listing, and restore through the same destination resolver. I’ll include focused backward-compatibility and command-construction tests and keep the submission scoped to the existing Destination model.

**Comment by themaksat** [comment]
I have submitted a solution for this issue in PR https://github.com/Dokploy/dokploy/pull/5493. Full implementation and verification included. Payout wallet (Solana): `2BTCUUTviLNDtZDRBhPBTwSXZMrTttPJNbGevMXBBMKK`

**Comment by prins1bap-ui** [comment]
/attempt #416

I’m taking a focused pass on the still-open Algora $50 bounty, targeting the current-main rclone path rather than introducing a parallel backup subsystem.

Plan:
- generalize the existing Destination abstraction so legacy S3 remains byte-for-byte compatible while a generic rclone remote/config can support Drive/OneDrive/FTP/SFTP;
- route connection test, upload, retention, listing, and restore through one validated helper instead of duplicating provider-specific path construction;
- fail closed on malformed remote/config input and extend secret redaction beyond the current S3-only patterns;
- add regression coverage for unchanged S3 behavior plus one generic-remote fixture across the backup lifecycle.

I’ll avoid overlapping the separate cloud-storage architecture in older PRs and keep the patch scoped to the existing backup pipeline.
